MapboxVision/src/main/java/com/mapbox/vision/utils/observable/ObserverComposer.kt

```diff
@@ -0,0 +1,36 @@
+package com.mapbox.vision.utils.observable
+
+import java.lang.ref.WeakReference
+import java.util.concurrent.CopyOnWriteArraySet
+
+open class ObserverComposer<T> : Observable<T> {
+    private val set = CopyOnWriteArraySet<WeakReference<T>>()
+    private var needClean = false
+
+    override fun addObservable(observer: T) {
+        set.add(WeakReference(observer))
+    }
+
+    override fun removeObserver(observer: T) {
+        set.find { it.get() == observer }?.let { set.remove(it) }
+    }
+
+    protected fun forEach(func: T.() -> Unit) {
+        set.forEach { weak ->
+            val instance = weak.get()
+            if (instance != null) {
+                instance.func()
+            } else {
+                needClean = true
```

Comment:
is it possible just to remove empty refs? without calling a separate method _cleanIfNeed()_
```suggestion
                set.remove(weak)
```

Reply:
hmmm, looks like we get OutOfBound Exception

Reply:
have you checked? as I remember we can do it with CopyOnWrite collections

Reply:
discussed, done